astropy/io/fits/file.py

```diff
@@ -487,22 +521,7 @@ def _open_filename(self, filename, mode, overwrite):
 
         ext = os.path.splitext(self.name)[1]
 
-        if ext == '.gz' or magic.startswith(GZIP_MAGIC):
-            # Handle gzip files
-            self._file = gzip.open(self.name, IO_FITS_MODES[mode])
-            self.compression = 'gzip'
-        elif ext == '.zip' or magic.startswith(PKZIP_MAGIC):
-            # Handle zip files
-            self._open_zipfile(self.name, mode)
-        elif ext == '.bz2' or magic.startswith(BZIP2_MAGIC):
-            # Handle bzip2 files
-            if mode in ['update', 'append']:
-                raise IOError("update and append modes are not supported "
-                              "with bzip2 files")
-            # bzip2 only supports 'w' and 'r' modes
-            bzip2_mode = 'w' if mode == 'ostream' else 'r'
-            self._file = bz2.BZ2File(self.name, bzip2_mode)
-        else:
+        if not self._test_for_compression(self.name, magic, mode, ext=ext):
```

Comment:
The method name is confusing, it doesn't only test, it also read the file if it is compressed. So maybe something like `_try_read_compressed` ?